### PR TITLE
Initial implementation of Angles in Paths

### LINKED ISSUE #490: Handle robot angles together with paths in a smart way
We've discussed having an "angle function" before that would be a property on the Path class.

cc @ashaw596

### PR DESCRIPTION
**Pull request by ashaw596**
Adds ability to add angle functions to paths as well as paths being able to return angle positions.

Closes #490.

### CODE REVIEW

File: soccer/planning/MotionInstant.hpp [hunk 1]
```diff
@@ -27,4 +28,32 @@ struct MotionInstant {
...
 
+/**
+ * This class represents a robot's angle "state" at a given time, including
+ * angle and rotation velocity.
+ */
+struct AngleInstant {
+    explicit AngleInstant(boost::optional<float> angle = boost::none,
+                          boost::optional<float> angleVel = boost::none)
+        : angle(angle), angleVel(angleVel){};
+    boost::optional<float> angle;
```
Comment (justbuchanan) [inline soccer/planning/MotionInstant.hpp]: Consider making these non-optional since AngleInstant is already optional inside of RobotInstant

File: soccer/Robot.hpp [hunk 2]
```diff
@@ -438,18 +439,17 @@ class OurRobot : public Robot {
...
-     * where mask values < 0 create no obstacle, and larger values
-     * create an obstacle of a given radius
-     *
-     * NOTE: mask must not be set for this robot
-     *
-     * @param robots is the set of robots to use to create a mask - either self
-     * or opp from _state
-     */
+    Planning::AngleFunctionPath angleFunctionPath;  /// latest path
+                                                    /**
```
Comment (justbuchanan) [inline soccer/Robot.hpp]: This comment is indented pretty strangely.  If you put a newline between the function and the variable it should fix it.

File: soccer/motion/MotionControl.cpp [hunk 3]
```diff
@@ -64,51 +76,47 @@ void MotionControl::run() {
...
     }
 
+    /*
+    _robot->addText(QString("targetW: %1").arg(targetW));
+    _robot->addText(QString("angleError: %1").arg(angleError));
+    _robot->addText(QString("targetGlobalAngle: %1").arg(targetAngleFinal));
+    _robot->addText(QString("angle: %1").arg(_robot->angle));
+    */
     _targetAngleVel(targetW);
+    //_targetAngleVel(targetW);
```
Comment (justbuchanan) [inline soccer/motion/MotionControl.cpp]: x

File: soccer/motion/MotionControl.cpp [hunk 4]
```diff
@@ -127,62 +135,45 @@ void MotionControl::run() {
...
+    _robot->state()->drawCircle(target.pos, .04, Qt::red, "MotionControl");
+    _robot->state()->drawLine(target.pos, target.pos + target.vel, Qt::blue,
+                              "MotionControl");
 
-        // convert from world to body coordinates
-        target.vel = target.vel.rotated(-_robot->angle);
-    }
+    // convert from world to body coordinates
+    target.vel = target.vel.rotated(-_robot->angle);
+    //    }
```
Comment (justbuchanan) [inline soccer/motion/MotionControl.cpp]: x

File: soccer/planning/MotionInstant.hpp [hunk 5]
```diff
@@ -16,10 +16,6 @@ struct MotionInstant {
     Geometry2d::Point pos;
     Geometry2d::Point vel;
 
-    friend bool operator==(const MotionInstant& a, const MotionInstant& b) {
```
Comment (justbuchanan) [inline soccer/planning/MotionInstant.hpp]: Any reason for removing `operator==()`?

File: soccer/planning/MotionInstant.hpp (hunk 1 above)
Comment (justbuchanan) [inline soccer/planning/MotionInstant.hpp]: @ashaw596 thoughts on this?

File: soccer/planning/Path.hpp [hunk 6]
```diff
@@ -19,17 +19,19 @@ class Path {
     Path(RJ::Time startTime = 0) : _startTime(startTime) {}
     virtual ~Path() {}
 
-    /**
-     * This method evalates the path at a given time and returns the target
-     * position and velocity of the robot.
+    /*
```
Comment (justbuchanan) [inline soccer/planning/Path.hpp]: This should be `/**` for doxygen

File: soccer/planning/Path.hpp [hunk 7]
```diff
@@ -87,11 +89,132 @@ class Path {
...
+public:
+    AngleFunctionPath(
+        std::unique_ptr<Path> path = nullptr,
+        boost::optional<std::function<AngleInstant(MotionInstant)>>
+            angleFunction = boost::none)
+        : path(std::move(path)), angleFunction(angleFunction) {}
+
+    std::unique_ptr<Path> path;
+    boost::optional<std::function<AngleInstant(MotionInstant)>> angleFunction;
+    /*
```
Comment (justbuchanan) [inline soccer/planning/Path.hpp]: Same here, `/**`.

File: soccer/planning/PivotPathPlanner.cpp [hunk 8]
```diff
@@ -0,0 +1,34 @@
...
+    return true;
+}
+
+std::unique_ptr<Path> PivotPathPlanner::run(
+    MotionInstant startInstant, const MotionCommand* cmd,
+    const MotionConstraints& motionConstraints,
+    const Geometry2d::ShapeSet* obstacles, std::unique_ptr<Path> prevPath) {
+    EscapeObstaclesPathPlanner escapePlanner;
+    EmptyCommand emptyCommand;
+    // TODO implement actual Pivoting
```
Comment (justbuchanan) [inline soccer/planning/PivotPathPlanner.cpp]: Maybe have this raise an exception for now until it's actually implemented?

Comment (justbuchanan) [inline soccer/planning/PivotPathPlanner.cpp]: Or even better, just remove it and put it in a PR that actually implements it

File: soccer/planning/RRTPlanner.cpp [hunk 9]
```diff
@@ -195,6 +195,28 @@ float getTime(InterpolatedPath& path, int index,
         motionConstraints.maxAcceleration, startSpeed, endSpeed);
 }
 
+std::unique_ptr<Planning::InterpolatedPath> RRTPlanner::generateVelocityPath(
```
Comment (justbuchanan) [inline soccer/planning/RRTPlanner.cpp]: This should probably also be in a different PR.

File: soccer/Robot.hpp (hunk 2 above)
Comment (ashaw596) [inline soccer/Robot.hpp]: fixed

File: soccer/motion/MotionControl.cpp (hunk 3 above)
Comment (ashaw596) [inline soccer/motion/MotionControl.cpp]: x

File: soccer/motion/MotionControl.cpp (hunk 4 above)
Comment (ashaw596) [inline soccer/motion/MotionControl.cpp]: x

File: soccer/planning/MotionInstant.hpp (hunk 5 above)
Comment (ashaw596) [inline soccer/planning/MotionInstant.hpp]: I didn't think we really needed it. Its not used anywhere and its not very useful for anything since you Point uses exact Point comparisons.

File: soccer/planning/Path.hpp (hunk 6 above)
Comment (ashaw596) [inline soccer/planning/Path.hpp]: fixed

File: soccer/planning/PivotPathPlanner.cpp (hunk 8 above)
Comment (ashaw596) [inline soccer/planning/PivotPathPlanner.cpp]: Fine, I'll raise an exception.

File: soccer/planning/RRTPlanner.cpp (hunk 9 above)
Comment (ashaw596) [inline soccer/planning/RRTPlanner.cpp]: Well, its a working implementation which just uses the current bezierPath method. (And I branched off this branch for my velocity stuff)

**Comment by ashaw596** [comment]
I Fixed most of the issues now I think.
